expression/builtin_control.go

```diff
@@ -50,47 +55,345 @@ type caseWhenFunctionClass struct {
+		decimal, flen, isBinaryStr = mathutil.Max(decimal, args[i].GetType().Decimal), mathutil.Max(flen, args[i].GetType().Flen), isBinaryStr || types.IsBinaryStr(args[i].GetType())
+	}
+	if l%2 == 1 {
+		fieldTps = append(fieldTps, args[l-1].GetType())
+		decimal, flen, isBinaryStr = mathutil.Max(decimal, args[l-1].GetType().Decimal), mathutil.Max(flen, args[l-1].GetType().Flen), isBinaryStr || types.IsBinaryStr(args[l-1].GetType())
+	}
+	fieldTp := types.AggFieldType(fieldTps)
+	classType := types.AggTypeClass(fieldTps, &fieldTp.Flag)
+	// Set retType to BINARY(0) if all arguments are of type NULL
+	if fieldTp.Tp == mysql.TypeNull {
+		fieldTp.Flen, fieldTp.Decimal = 0, 0
+		types.SetBinChsClnFlag(fieldTp)
+	}
+	fieldTp.Decimal, fieldTp.Flen = decimal, flen
+	var tp evalTp
+	switch classType {
+	case types.ClassInt:
+		tp = tpInt
+		fieldTp.Decimal = 0
+	case types.ClassReal:
+		tp = tpReal
+	case types.ClassDecimal:
+		tp = tpDecimal
+	case types.ClassString:
+		tp = tpString
+		if !isBinaryStr {
+			fieldTp.Charset, fieldTp.Collate = mysql.DefaultCharset, mysql.DefaultCollationName
+		}
+		if types.IsTypeTime(fieldTp.Tp) {
+			tp = tpTime
+		} else if fieldTp.Tp == mysql.TypeDuration {
+			tp = tpDuration
+		}
+	}
+
+	argTps := make([]evalTp, 0, l)
+	for i := 0; i < l-1; i += 2 {
+		argTps = append(argTps, tpInt, tp)
+	}
+	if l%2 == 1 {
+		argTps = append(argTps, tp)
+	}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tp, argTps...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf.tp = fieldTp
+	switch classType {
+	case types.ClassInt:
+		sig = &builtinCaseWhenIntSig{baseIntBuiltinFunc{bf}}
+	case types.ClassReal:
+		sig = &builtinCaseWhenRealSig{baseRealBuiltinFunc{bf}}
+	case types.ClassDecimal:
+		sig = &builtinCaseWhenDecimalSig{baseDecimalBuiltinFunc{bf}}
+	case types.ClassString:
+		sig = &builtinCaseWhenStringSig{baseStringBuiltinFunc{bf}}
+		if types.IsTypeTime(fieldTp.Tp) {
+			sig = &builtinCaseWhenTimeSig{baseTimeBuiltinFunc{bf}}
+		} else if fieldTp.Tp == mysql.TypeDuration {
+			sig = &builtinCaseWhenDurationSig{baseDurationBuiltinFunc{bf}}
+		}
+	}
 	return sig.setSelf(sig), nil
 }
 
-type builtinCaseWhenSig struct {
-	baseBuiltinFunc
+type builtinCaseWhenIntSig struct {
+	baseIntBuiltinFunc
 }
 
-// eval evals a builtinCaseWhenSig.
+// evalInt evals a builtinCaseWhenIntSig.
 // See https://dev.mysql.com/doc/refman/5.7/en/case.html
-func (b *builtinCaseWhenSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return types.Datum{}, errors.Trace(err)
+func (b *builtinCaseWhenIntSig) evalInt(row []types.Datum) (int64, bool, error) {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	var (
+		arg    int64
+		isNull bool
+		err    error
+	)
+	args, l := b.getArgs(), len(b.getArgs())
+	for i := 0; i < l-1; i += 2 {
+		arg, isNull, err = args[i].EvalInt(row, sc)
+		if err != nil {
+			return 0, false, errors.Trace(err)
```

Comment:
I think it doesn't matter whether `isNull == true` or `isNull == false` when `err != nil` here